feat: add support for batch execution in parallel with custom Executor

visheshruparelia opened the pull request:
#1864

## Description of changes:

Add a new method in the BatchHandler class which allows users to pass their custom Executor for processing messages in parallel.

**Checklist**

<!--- Leave unchecked if your change doesn't seem to apply --> 

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda-java/#tenets)
* [x] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/)

## Breaking change checklist

<!--- Ignore if it's not a breaking change -->

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

visheshruparelia commented on powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/internal/MultiThreadMDC.java:
```diff
@@ -44,4 +44,11 @@ public void copyMDCToThread(String thread) {
         }
     }
+
+    public void removeThread(String thread) {
```
I think this will be helpful if the same invocation of `processBatchInParallel()` ends up re-using the thread.

phipag replied:
Yes, this is good. I believe it was a small memory leak because we never removed the thread either.

phipag commented on powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/KinesisStreamsBatchMessageHandler.java:
```diff
@@ -77,7 +81,9 @@ public StreamsEventResponse processBatchInParallel(KinesisEvent event, Context c
                     multiThreadMDC.copyMDCToThread(Thread.currentThread().getName());
-                    return processBatchItem(eventRecord, context);
+                    Optional<StreamsEventResponse.BatchItemFailure> failureOpt = processBatchItem(eventRecord, context);
+                    multiThreadMDC.removeThread(Thread.currentThread().getName());
```
Good catch!!

phipag commented:
Hey @visheshruparelia, I finally reviewed your PR and it works great. I just tested it end-to-end using the examples you also added. Thanks for this! 🚀 

Can you please have a look at the pmd_analyse comments in the diff and address them? I added comments to some but there are more such as:

> A catch statement should never catch throwable since it includes errors.
> Catching Throwable errors is not recommended since its scope is very broad. It includes runtime issues such as
OutOfMemoryError that should be exposed and managed separately.
> AvoidCatchingThrowable (Priority: 1, Ruleset: Error Prone)
https://docs.pmd-code.org/snapshot/pmd_rules_java_errorprone.html#avoidcatchingthrowable

I think we can just catch `Exception` there.

Regarding the code duplication reported by Sonarcube, I think it is okay in this case since it is actually easier to read for developers than creating more abstractions in addition to the existing ones.

One last thing is the documentation page [here](https://github.com/aws-powertools/powertools-lambda-java/blob/1fa11c138529ee7d49f6d24415a5de6223330aa2/docs/utilities/batch.md). Do you mind documenting the new feature that you implemented there? You could add a second example showing how to pass a custom executor to this code snippet as a second tab:

![image](https://github.com/user-attachments/assets/8b8b53c2-82c9-43e7-8d53-bc118fd7fe5e)

phipag commented:
> > ## [![Quality Gate Failed](https://camo.githubusercontent.com/ce402bcc4cf818eda660128f0b27a891eb8c9f4e0d783303ea667cb0d4b07f95/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f71672d6661696c65642d323070782e706e67)](https://sonarcloud.io/dashboard?id=aws-powertools_powertools-lambda-java&pullRequest=1900) **Quality Gate failed**
> > Failed conditions ![](https://camo.githubusercontent.com/bd1d5234b37410841ad8cd9d85c84eeecaec92ae18a95c926ae3a59a17c29a65/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6661696c65642d313670782e706e67) [9.8% Duplication on New Code](https://sonarcloud.io/component_measures?id=aws-powertools_powertools-lambda-java&pullRequest=1900&metric=new_duplicated_lines_density&view=list) (required ≤ 3%)
> > [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=aws-powertools_powertools-lambda-java&pullRequest=1900)
> 
> How come did the code duplication increase so much? Is this old code that is now being scanned, or new code? Can we do anything to improve it?

The code duplication was there before and triggered again as a finding in this PR due to implementing the new method for `processBatchInParallel`:


```java
    @Override
    public StreamsEventResponse processBatchInParallel(KinesisEvent event, Context context) {
        MultiThreadMDC multiThreadMDC = new MultiThreadMDC();

        List<StreamsEventResponse.BatchItemFailure> batchItemFailures = event.getRecords()
                .parallelStream() // Parallel processing
                .map(eventRecord -> {
                    multiThreadMDC.copyMDCToThread(Thread.currentThread().getName());
                    Optional<StreamsEventResponse.BatchItemFailure> failureOpt = processBatchItem(eventRecord, context);
                    multiThreadMDC.removeThread(Thread.currentThread().getName());
                    return failureOpt;
                })
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toList());

        return StreamsEventResponse.builder().withBatchItemFailures(batchItemFailures).build();
    }
```

This code snippet is essentially the same for DynamoDB and Kinesis (code duplication reported here) – but it is different for SQS because we need to check if we are dealing with a FIFO queue. It would require some larger refactoring if we wanted to remove the code duplication completely. 

My personal opinion is, that we should favor the current code duplication because:
- It is more readable (we don't have to jump between Abstract class **and** interface implementations)
- It is more flexible if we need to make adjustments for details of a specific event source (e.g. SQS FIFO)
- It is easier for developers to add a new implement. Just implement the `BatchMessageHandler` interface. I don't need to refactor again some abstracted logic

If we think that we should refactor though, I am also okay with that. Let me know what you folks think @dreamorosi and @visheshruparelia .

dreamorosi commented:
> This code snippet is essentially the same for DynamoDB and Kinesis (code duplication reported here) – but it is different for SQS because we need to check if we are dealing with a FIFO queue. It would require some larger refactoring if we wanted to remove the code duplication completely.
> 
> My personal opinion is, that we should favor the current code duplication because:


Sounds sensible, thanks for explaining.

visheshruparelia commented:
Hi @phipag, thank you for the review. 

>Can you please have a look at the pmd_analyse comments in the diff and address them?
Most are on the code which is untouched in this PR. Would you prefer combine unrelated refactoring + the main code changes in the same PR?

phipag commented on examples/powertools-examples-batch/src/main/java/org/demo/batch/dynamo/DynamoDBStreamBatchHandlerParallel.java:
```diff
@@ -0,0 +1,37 @@
+        return handler.processBatchInParallel(ddbEvent, context, executor);
+    }
+
+    private void processMessage(DynamodbEvent.DynamodbStreamRecord dynamodbStreamRecord, Context context) {
```
Thanks, this would be awesome if you can update those as well. Ideally, we have no Sonar or pmd_analyze findings before merging the PR. With the exception of the code duplication finding.

visheshruparelia commented:
> Hey @visheshruparelia, I finally reviewed your PR and it works great. I just tested it end-to-end using the examples you also added. Thanks for this! 🚀
> 
> Can you please have a look at the pmd_analyse comments in the diff and address them? I added comments to some but there are more such as:
> 
> > A catch statement should never catch throwable since it includes errors.
> > Catching Throwable errors is not recommended since its scope is very broad. It includes runtime issues such as
> > OutOfMemoryError that should be exposed and managed separately.
> > AvoidCatchingThrowable (Priority: 1, Ruleset: Error Prone)
> > https://docs.pmd-code.org/snapshot/pmd_rules_java_errorprone.html#avoidcatchingthrowable
> 
> I think we can just catch `Exception` there.
> 
> Regarding the code duplication reported by Sonarcube, I think it is okay in this case since it is actually easier to read for developers than creating more abstractions in addition to the existing ones.
> 
> One last thing is the documentation page [here](https://github.com/aws-powertools/powertools-lambda-java/blob/1fa11c138529ee7d49f6d24415a5de6223330aa2/docs/utilities/batch.md). Do you mind documenting the new feature that you implemented there? You could add a second example showing how to pass a custom executor to this code snippet as a second tab:
> 
> ![image](https://private-user-images.githubusercontent.com/5087450/464601536-8b8b53c2-82c9-43e7-8d53-bc118fd7fe5e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIyNDQ4MTEsIm5iZiI6MTc1MjI0NDUxMSwicGF0aCI6Ii81MDg3NDUwLzQ2NDYwMTUzNi04YjhiNTNjMi04MmM5LTQzZTctOGQ1My1iYzExOGZkN2ZlNWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MTFUMTQzNTExWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTgzY2QzNDdkNTFkMjkwY2ZlNDFkY2M5MDI4Y2QzNmY4ZGEyZjQ5YWU2Y2VlYzliZWNhYjYyOWM1ZmE5NjAzMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.gQb6DGJnH8qnYOQ8o-Xn9eilixh_cyYMxb_YfDwuwaY)

4d0cd06

phipag commented on docs/utilities/batch.md:
````diff
@@ -510,6 +510,31 @@ used with SQS FIFO. In that case, an `UnsupportedOperationException` is thrown.
     ```
+=== "Example with SQS (using custom executor)"
+
+    ```java hl_lines="13"
````
```suggestion
    ```java hl_lines="4 10 15"
```

visheshruparelia commented:
@phipag done
